docs/design/drf - fairshare.md

```diff
@@ -0,0 +1,179 @@
+> 
+> After the scheduling of job1, right order should be job3-job4-job2. But in priority queue, we have no chance to fix the priority for job2
+
+#### Namespace Order
+
+To add namespace weight, we introduce a new order function named `NamespaceOrder` in `Session`.
+
+```go
+type Session struct {
+    ...
+    NamespaceOrderFns map[string]api.CompareFn
+    ...
+}
+```
+
+The scheduling loop in allocate should change as follows.
+
+In scheduling loop, firstly, choose a namespace having highest priority by calling `NamespaceOrderFn`, and then choose a job having highest priority using `JobOrderFn` in this namespace.
+
+After scheduling of job, push the namespace and job back to priority queue in order to refresh its priority. Because once a job is scheduled, assigned resource may decrease the priority of this namespace, the other jobs in the same namespace may be scheduled later.
+
+Always assigning resources to namespace with highest priority (lower resource usage) in every turn will make the resource balanced.
+
+### DRF plugin
+
+DRF plugin use preemption and order of job to balance resource among jobs. The `share` in this plugin is defined as resource usage, the higher `share` means this job occupies the more resource now.
+
+#### Namespace Compare
+
+To introduce namespace weight into this plugin, we should define how to compare namespace having weight firstly.
+
+For namespace n1 having weight w1 and namespace n2 having weight w2, we can compute the `share` of resource and recorded as u1 and u2. Now, the resource usage of n1 less than n2 can be defined as (u1 / w1 < u2 / w2)
+
+`e.g.` ns1 having weight w1=2 use 6cpu, ns2 having weight w2=1 use 2cpu. In the scope of cpu, the ns1 use less resource than ns2. (6 / 3 < 3 / 1)
+
+#### Namespace Order
+
+Register `NamespaceOrder` function using the strategy mentioned above.
+
+#### preemption
+
+> The `preempt` action is disabled now. Do this later.
+
+In the `preemption` function now, strategy is just simply comparing the resource share among jobs .
+
+After adding namespace weight, we should check namespace of preemptor and preemptee firstly. The job in namespace with less resources can preempt others, or when namespace resource usage are the same, compare share of job instead.
+
+### Feature Interaction
+
+#### preempt action
+
+Preempt is a strategy set to choose victims and finally evict it.
+
+The way to choose victims is a function set named `Preemptable` registered by plugins. Such as job ready protection of Gang plugin, special pod protection of Conformance plugin, job share balance strategy of DRF plugin.
+
+All these plugin would choose some victims respective, and the intersection of them would be the final victim set. So, the choice made by DRF plugin would never break the requirement of others.
+
+### short hand
+
+1. Preempt may cause killing of some running pod.
+
+### Cases:
+
+- cluster have __16 cpu__, queue and namespace have default weight.
```

Comment:
When namespace requests more resources (calculated from pods) than it's resources quota(from v1.ResourceQuota). At the same time, this namespace has higher weight than the other namespaces in the same queue? Is it a possibility some resources allocated by volcano but failed resource quota check?

Reply:
For this case, the request of more resources than it's resources quota will cause error like this 

`pods \"lm-fairshare-wnchr-tb-3\" is forbidden: exceeded quota: compute-resources, requested: limits.cpu=500m, used: limits.cpu=2, limited: limits.cpu=2`

So, if that pod is `Pending` and wait for allocation from volcano, it is already satisfied the resource quota.